solc/CommandLineInterface.cpp

```diff
@@ -1776,14 +1776,20 @@ bool CommandLineInterface::link()
 				return false;
 			}
 
-			string name(it, it + placeholderSize);
-			if (librariesReplacements.count(name))
+			if (*(it + 1) != '_' || *(it + placeholderSize - 2) != '_' || *(it + placeholderSize - 1) != '_')
```

Comment:
Is there an assumption on the bounds of `placeholderSize`?

Reply:
It must be exactly 40 bytes. I guess the constant is here not because it can change but to give it a name and to avoid hard-coding it. It's officially documented as 40 bytes and that's enough to store a hash so I don't think it will ever change.

Reply:
Why do we have this messy code here and why isn't this part of `LinkerObject`? Also `solc-js` has this messy code duplicated.

Reply:
Ah this is the feature for linking previously unlinked stuff, so we don't have the `LinkerObject`. However I would suggest to check out `solc-js` and make sure the logic there and here matches.

Reply:
I asked myself the same question when I first saw it was separate from `LinkerObject` :) I even have a half-done rewrite branch but I ultimately stopped because I realized I would have to make it actually parse the input file. Right now it just rolls over everything, both binary and the text after it. It also does not decode it from hex so it works even with invalid hex encoding as long as link references are fine.

I can finish the refactor and submit a PR if you think it's worthwhile.

Reply:
@axic
> Ah this is the feature for linking previously unlinked stuff, so we don't have the LinkerObject.

Yeah, but it's pretty easy to create one. The only obstacle is that you need to separate the bytecode from the other stuff in the file.

Reply:
I would prefer a clear way of adding a `LinkerObject::parse(bytes)` feature and unit testing and using `parse().link(addresses)` in the cli.

Reply:
That's pretty much what I was trying to do with my refactor (just `fromHex()` instead of `parse()`).

In that case I'll finish it and submit a PR. But it'll be separate from this one.

Reply:
You can take inspiration from https://github.com/ethereum/solc-js/blob/master/linker.js (see `findLinkReferences`).

Reply:
ok. I have a heap of review comments and also other tasks to get through first so I created an issue so that it does not get forgotten: #10308.